Make DB interval adjustable

DL6ER opened the pull request:
**By submitting this pull request, I confirm the following (please check boxes, eg [X]) _Failure to fill the template will close your PR_:**

***Please submit all pull requests against the `development` branch. Failure to do so will delay or deny your request***

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/FTL/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:** 

## 10

---

Make interval in which FTL stores queries in the long term database easily adjustable via the config file. 

It defaults to `1.0` (once per minute).
It cannot be lower than `0.1` (every six seconds).
It cannot be larger than `1440.0` (once a day).

Related to https://github.com/pi-hole/pi-hole/issues/1839

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

Mcat12 commented:
Wouldn't writing to the database in intervals longer than a day miss some queries?

Mcat12 commented on config.c:
```diff
@@ -130,6 +130,26 @@ void read_FTLconf(void)
+	// How often do we store queries in FTL's database [minutes]?
+	// this value can be a floating point number, e.g. "DBINTERVAL=0.5"
+	// defaults to: 1 (once per minute)
+	config.DBinterval = 1;
```
The default should be 60 (once every 60 seconds) based on how it is used below.

graysky2 commented:
@DL6ER - I'm thinking that you probably also want to force a write-out when `/usr/bin/pihole-FTL` exits?

DL6ER commented:
> force a write-out when /usr/bin/pihole-FTL exits

No, I don't really want that. It is unforeseeable how long the write will take and I consider it as bad practice to react only delayed to requests to terminate. I think a sane write out interval (i.e. several times an hour) should be sufficient to achieve that everything is properly saved. 

There is only one scenario where you could loose some queries and that is having a very long interval, shutting down your Pi-hole and then powering it back on at least 48 hours later. I'm not sure if that is a very likely scenario as you wouldn't have DNS functionality in your network as long as Pi-hole is shut down.

graysky2 commented:
@DL6ER - Sounds good, just wanted to check.

graysky2 commented:
@DL6ER - The [Arch package](https://aur.archlinux.org/cgit/aur.git/commit/?h=pi-hole-ftl&id=6efb21f2818cc18b68e86cbf03e79fe742c6f510) for pi-hole-ftl has this PR#172 and PR#167 applied to the standard v2.12.  I have `/etc/pihole/pihole-FTL.conf` setup to sync every 121 minutes, but I haven't seen a sync occur (4 hours of uptime).

Local time is 5:25 PM:
```
% ls -lrt /etc/pihole
...
-rw-r--r-- 1 pihole pihole  147456 Dec 13 12:30 pihole-FTL.db
```

```
% cat /etc/pihole/pihole-FTL.conf 
SOCKET_LISTENING=localonly
TIMEFRAME=rolling24h
QUERY_DISPLAY=yes
AAAA_QUERY_ANALYSIS=yes
MAXDBDAYS=365
RESOLVE_IPV6=no
RESOLVE_IPV4=yes
DBINTERVAL=121.0
DBFILE=/etc/pihole/pihole-FTL.db
```

DL6ER commented:
Strange... I will set the timer to 121.0 for me as well and report tomorrow what happened over night.

graysky2 commented:
I got the same...
```
% grep -i dbint /run/log/pihole-ftl/pihole-FTL.log 
[2017-12-13 13:26:21.132]    DBINTERVAL: saving to DB file every 7260 seconds
```

Yet still no writes...
```
% stat /etc/pihole/pihole-FTL.db 
  File: /etc/pihole/pihole-FTL.db
  Size: 147456    	Blocks: 288        IO Block: 4096   regular file
Device: 30h/48d	Inode: 260690      Links: 1
Access: (0644/-rw-r--r--)  Uid: (  996/  pihole)   Gid: (  996/  pihole)
Access: 2017-12-13 14:49:00.053508686 -0500
Modify: 2017-12-13 12:30:00.196199782 -0500
Change: 2017-12-13 13:26:21.298180015 -0500
 Birth: -
```

DL6ER commented:
@graysky2 It seems to work fine for me, writes are done every 02:01 (HH:MM):
```
[2017-12-14 00:51:00.185] Notice: Queries stored in DB: 826
[2017-12-14 00:51:00.186] Notice: Database size is 138.87 MB, deleted 0 rows
[...]
[2017-12-14 02:52:00.192] Notice: Queries stored in DB: 382
[2017-12-14 02:52:00.194] Notice: Database size is 138.90 MB, deleted 0 rows
[...]
[2017-12-14 04:53:00.184] Notice: Queries stored in DB: 412
[2017-12-14 04:53:00.185] Notice: Database size is 138.92 MB, deleted 0 rows
[...]
[2017-12-14 06:54:00.143] Notice: Queries stored in DB: 391
[2017-12-14 06:54:00.145] Notice: Database size is 138.95 MB, deleted 0 rows
```

graysky2 commented:
Are you using 2,12 plus those 2 patches or the dev branch?

On Thursday, December 14, 2017, DL6ER <notifications@github.com> wrote:

> @graysky2 <https://github.com/graysky2> It seems to work fine for me,
> writes are done every 02:01 (HH:MM):
>
> [2017-12-14 00:51:00.185] Notice: Queries stored in DB: 826
> [2017-12-14 00:51:00.186] Notice: Database size is 138.87 MB, deleted 0 rows
> [...]
> [2017-12-14 02:52:00.192] Notice: Queries stored in DB: 382
> [2017-12-14 02:52:00.194] Notice: Database size is 138.90 MB, deleted 0 rows
> [...]
> [2017-12-14 04:53:00.184] Notice: Queries stored in DB: 412
> [2017-12-14 04:53:00.185] Notice: Database size is 138.92 MB, deleted 0 rows
> [...]
> [2017-12-14 06:54:00.143] Notice: Queries stored in DB: 391
> [2017-12-14 06:54:00.145] Notice: Database size is 138.95 MB, deleted 0 rows
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pi-hole/FTL/pull/172#issuecomment-351622962>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAgpJpuOZwbEObH9TZk_5ZfpeqNfL5IPks5tAMNngaJpZM4Q_Zna>
> .
>


-- 
Sent from Gmail Mobile

DL6ER commented:
I'm on the branch of this PR. I quickly reviewed what that means:

`v2.12 + #166 + #168 + #171 + all comits in this PR`

However, nothing outside of this PR should be changing anything here. How about other values? I think you said already that `60` is working, how about `61`?

graysky2 commented:
Very odd... a value of 2 works just fine.  I am testing 61 now and will report back.

graysky2 commented:
Nope... a value of 61 has failed to write.

```
% grep -i dbint /run/log/pihole-ftl/pihole-FTL.log
[2017-12-14 14:51:19.538]    DBINTERVAL: saving to DB file every 3660 seconds
```
Local time is 16:55:
```
% stat /etc/pihole/pihole-FTL.db
  File: /etc/pihole/pihole-FTL.db
  Size: 208896    	Blocks: 408        IO Block: 4096   regular file
Device: 30h/48d	Inode: 260690      Links: 1
Access: (0644/-rw-r--r--)  Uid: (  996/  pihole)   Gid: (  996/  pihole)
Access: 2017-12-14 14:51:19.558958997 -0500
Modify: 2017-12-14 14:50:00.107371181 -0500
Change: 2017-12-14 14:50:00.107371181 -0500
 Birth: -
```

DL6ER commented:
@graysky2 I'm confused. Are you sure that your Arch version hast the whole PR merged into it? Including 5951e5b ?

graysky2 commented:
@DL6ER - Looks like that is omitted... see here: https://aur.archlinux.org/cgit/aur.git/commit/?h=pi-hole-ftl&id=6efb21f2818cc18b68e86cbf03e79fe742c6f510

graysky2 commented:
So our package should include 2.12 +
  https://github.com/pi-hole/FTL/commit/95bdd62777347f2117070ce8060b2ca1cc3072c0
  https://github.com/pi-hole/FTL/commit/effca0465fe9fc71500dd28ef31583902bdc9754
  https://github.com/pi-hole/FTL/commit/5951e5bde57ef8e22c5b191abdc4edc0299057fb
  https://github.com/pi-hole/FTL/commit/8ab56e58f29fc3e5263ab51b80534154832ffbd6
  https://github.com/pi-hole/FTL/commit/5951e5bde57ef8e22c5b191abdc4edc0299057fb

Correct?

Mcat12 commented:
Are you trying to make the current branch a package? Why do you need to distinguish between commits?

graysky2 commented:
@Mcat12 - No, the intent is incorporate this functionality (an adjustable db) into 2.12.  The patch I referenced above is working with a `DBINTERVAL=121.0`

DL6ER commented:
Okay, so there is no issue with this PR or our code base but only in the way you applied which modifications to the last tagged release. Good to know, thanks for doing extensive testing BTW.

graysky2 commented:
@DL6ER - You're welcome and thank you for the quick implementation of this request.  Is the patch I referenced above the correct set of commits to apply to 2.12 for the Arch package?

DL6ER commented:
@graysky2 

> Is the patch I referenced above the correct set of commits to apply to 2.12 for the Arch package?

Not sure what/which link you are referring to...

graysky2 commented:
Thanks again.  Will wait for 2.12.1 or whatever you bump to.
